src/otaclient/_otaproxy_ctx.py

```diff
@@ -39,26 +39,23 @@
 logger = logging.getLogger(__name__)
 
 _otaproxy_p: mp_ctx.SpawnProcess | None = None
-_global_shutdown: bool = False
-_global_shutdown_lock = threading.Lock()
+_global_shutdown = threading.Event()
+_global_process_lock = threading.Lock()
 
 
 def otaproxy_on_global_shutdown() -> None:
     global _global_shutdown
-    _global_shutdown = True
+    _global_shutdown.set()
     _shutdown_otaproxy()
 
 
 def _shutdown_otaproxy():
     global _otaproxy_p
-    if _global_shutdown_lock.acquire(blocking=False):
-        try:
-            if _otaproxy_p:
-                _otaproxy_p.terminate()
-                _otaproxy_p.join()
-                _otaproxy_p = None
-        finally:
-            _global_shutdown_lock.release()
+    with _global_process_lock:
+        if _otaproxy_p:
+            _otaproxy_p.terminate()
+            _otaproxy_p.join()
+            _otaproxy_p = None
```

Comment:
Please use blocking=False here, this is on purpose.
Originally the shutdown handler is for signal handling, when multiple signals arrived, we only need one signal actually being handled at a time, while not blocking other signals handling(directly return).

For example, if we issue multiple `kill` at the same time, if we using blocking lock in the handler, all the `kill` except one will be blocked, that is not needed.

Reply:
We can add an option `blocking` for the `_shutdown_otaproxy` function.

Reply:
Ah sorry I misunderstood that this handler is the signal handler, but it is not. Then your original changes are OK. :+1: